feat: #15416 Expand round robin availability beyond default schedule

vijayraghav-io opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

- Fixes #15416 
- Fixes CAL-3926

/claim #15416 

@ciaranha @itsberkaya 

https://www.loom.com/share/b194ce55a54649edad65eb10587994c6?sid=3faf6346-e165-4e4d-adf4-a4a9f0193bb7

<!-- Please provide a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A -  I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

vijayraghav-io committed:
add feature to change host schedule

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#15416_roundRobinAvailability

anikdhabal commented:
Hey @vijayraghav-io, great work! However, this issue needs a design, and we are waiting for the designer's approval or a new design. Make this PR a draft until we receive their input.

Thanks for your work.

Ryukemeister commented:
@vijayraghav-io I know this PR is still in the draft phase, but can you have a look at why type check are failing?

vijayraghav-io commented:
Yes checking

vijayraghav-io committed:
update for type-check:ci

vijayraghav-io committed:
Update SingleForm.tsx for linterr

vijayraghav-io commented:
> @vijayraghav-io I know this PR is still in the draft phase, but can you have a look at why type check are failing?

updated for errors, now tests are passing

vijayraghav-io committed:
update as per design review

vijayraghav-io committed:
update for rerendering

Amit91848 reviewed (changes requested):
Seems to me like you forgot to include migration file, please add those changes @vijayraghav-io 🙏

Amit91848 committed:
Merge branch 'main' into feat_#15416_roundRobinAvailability

Amit91848 commented on packages/trpc/server/routers/viewer/availability/schedule/getAllSchedulesByUserId.handler.ts:
```diff
@@ -0,0 +1,64 @@
+
+  if (!isCurrentUserPartOfTeam && !isCurrentUserOwner) {
+    throw new TRPCError({
+      code: "UNAUTHORIZED",
```
Check for user role before querying the database

vijayraghav-io replied:
Here the checks  `if(!isCurrentUserPartOfTeam && !isCurrentUserOwner)` , ensures that data (i.e. availabilities) of current user's or his team member's only can be queried. It should be possible regardless of the role.

Amit91848 replied:
I meant do this check `if(!isCurrentUserPartOfTeam && !isCurrentUserOwner)`  before querying the database for schedules

vijayraghav-io replied:
oh, ok! done

Amit91848 reviewed (changes requested):
Left some comments

vijayraghav-io committed:
migration files

vijayraghav-io committed:
Merge branch 'feat_#15416_roundRobinAvailability' of https://github.com/vijayraghav-io/cal.com into feat_#15416_roundRobinAvailability

vijayraghav-io committed:
On Delete set null for schedule

if schedule is deleted, no need to delete the host

vijayraghav-io committed:
updated schema

vijayraghav-io committed:
added testcases

vijayraghav-io commented:
> Seems to me like you forgot to include migration file, please add those changes @vijayraghav-io 🙏

Added migration file

vijayraghav-io commented on apps/web/components/eventtype/EventAvailabilityTab.tsx:
```diff
@@ -278,25 +279,168 @@ const EventTypeSchedule = ({ eventType }: { eventType: EventTypeSetup }) => {
+  const hosts = getValues("hosts");
+
+  const getIdx = () =>
+    Math.max(
```
Its just that selection is defaulted to first element in available options by setting to 0, no wrong data is sent to backend.

vijayraghav-io committed:
checking permissions before

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#15416_roundRobinAvailability

vijayraghav-io committed:
Merge branch 'main' into feat_#15416_roundRobinAvailability

vijayraghav-io committed:
Merge branch 'feat_#15416_roundRobinAvailability' of https://github.com/vijayraghav-io/cal.com into feat_#15416_roundRobinAvailability

Amit91848 committed:
Merge branch 'main' into feat_#15416_roundRobinAvailability

Amit91848 commented on apps/web/modules/event-types/views/event-types-single-view.tsx:
```diff
@@ -119,7 +119,15 @@ const ManagedEventTypeDialog = dynamic(() => import("@components/eventtype/Manag
+  userId: number;
+  priority: number;
+  scheduleId: number | null;
+  availability?: AvailabilityOption | null;
```
Why are you storing AvailabilityOption? We are already fetching all schedules of the user, you can find selected Option by matching it against scheduleId.

vijayraghav-io replied:
This is required to remember the user selection for scenario when user is toggling between `Choose a common schedule`. When common schedule is choosen scheduleId is made null.

Amit91848 commented on apps/web/components/eventtype/EventAvailabilityTab.tsx:
```diff
@@ -278,25 +279,168 @@ const EventTypeSchedule = ({ eventType }: { eventType: EventTypeSetup }) => {
+                      <Avatar size="sm" imageSrc={host.avatar} alt={host.label || ""} />
+                      <p className="text-emphasis my-auto ms-3 text-sm">{host.label}</p>
+                    </div>
+                    <TeamMemberSchedule host={host} />
```
Pass the index as props instead of having to search it using getIdx.

vijayraghav-io replied:
was more cautious , as the hosts index is changed when priority in TeamTab (or Assignment Tab) is changed due to sorting. Since TeamMemberSchedule is child component of TeamAvailability, this is not an issue, tested and updated.

Amit91848 commented on apps/web/modules/event-types/views/event-types-single-view.tsx:
```diff
@@ -119,7 +119,15 @@ const ManagedEventTypeDialog = dynamic(() => import("@components/eventtype/Manag
+  priority: number;
+  scheduleId: number | null;
+  availability?: AvailabilityOption | null;
+  avatar?: string;
+  label?: string;
```
I don't think we should extend host to include avatar and label especially since we are not storing it. You can use getValue and find avatar and label of the host inside TeamMemberSchedule.

vijayraghav-io replied:
This is only for enriching the fetched data, avatar and label are not stored through hosts, these fields are removed before sending payload.
Also wanted to be consistent with Assignment Tab. The label in EventTeamTab (Assignment) is fetched from EventSetupProps.teamMembers. This prop in not available in AvailabilityTab, else requires similar changes.

Amit91848 replied:
My bad, I meant you can access team members from eventType that is passed as props to UseCommonScheduleSettings by eventType.team.members[number].user and get avatar and label from there. 

Extending hosts is unnecessary as then you would have to clean it up later, best use eventType which is already being passed.

Amit91848 commented on apps/web/components/eventtype/EventAvailabilityTab.tsx:
```diff
@@ -278,25 +279,168 @@ const EventTypeSchedule = ({ eventType }: { eventType: EventTypeSetup }) => {
+  }, [data]);
+
+  const idx = getIdx();
+  const availabilityValue = watch(`hosts.${idx}.availability`);
```
No need to store the availabilityValue, you can find the selected option using scheduleId and schedules.

vijayraghav-io replied:
used to set initial value, watch is not required. Updated.

Amit91848 commented on apps/web/components/eventtype/EventAvailabilityTab.tsx:
```diff
@@ -278,25 +279,168 @@ const EventTypeSchedule = ({ eventType }: { eventType: EventTypeSetup }) => {
+        ? option.value === scheduleId
+        : availability
+        ? option.value === availability?.value
+        : option.value === schedules.find((schedule) => schedule.isDefault)?.id
```
This won't be necessary now. since we are only using scheduleId.

vijayraghav-io replied:
This is required to set default values when new host is added

Amit91848 reviewed (changes requested):
Great work so far @vijayraghav-io . Left a few comments. Lmk your views.

Also when you add a host, it resets the availability of host of the same type, i.e if you add a fixed host, it changes the availability of other fixed host to default and same with round robin hosts.


https://github.com/user-attachments/assets/83935797-2988-49c9-a1a5-22f0320b1b29

vijayraghav-io commented:
> Great work so far @vijayraghav-io . Left a few comments. Lmk your views.
> 
> Also when you add a host, it resets the availability of host of the same type, i.e if you add a fixed host, it changes the availability of other fixed host to default and same with round robin hosts.
> 
>  Recording.2024-07-15.160752.mp4

Thankyou @Amit91848 , i am going through the comments and will address them.

vijayraghav-io committed:
updated for comments

vijayraghav-io committed:
Merge branch 'feat_#15416_roundRobinAvailability' of https://github.com/vijayraghav-io/cal.com into feat_#15416_roundRobinAvailability

vijayraghav-io commented:
> Great work so far @vijayraghav-io . Left a few comments. Lmk your views.
> 
> Also when you add a host, it resets the availability of host of the same type, i.e if you add a fixed host, it changes the availability of other fixed host to default and same with round robin hosts.
> 
>  Recording.2024-07-15.160752.mp4

I have updated fix for this and is working fine for same hosts and different hosts. My bad i was checking for different kind of hosts all the time. Great find!

vijayraghav-io commented on apps/web/components/eventtype/EventAvailabilityTab.tsx:
```diff
@@ -278,25 +279,168 @@ const EventTypeSchedule = ({ eventType }: { eventType: EventTypeSetup }) => {
+      isManaged: false,
+    }));
+
+    setOptions(scheduleOptions);
```
Yeah useEffect is not required, i had started with functionality of showing availability in TeamTab that needed useEffect(for sorting on priority), its not required now, removed.

vijayraghav-io commented on apps/web/components/eventtype/EventAvailabilityTab.tsx:
```diff
@@ -278,25 +279,168 @@ const EventTypeSchedule = ({ eventType }: { eventType: EventTypeSetup }) => {
+  });
+
+  useEffect(() => {
+    if (!data || isPending) {
```
updated

vijayraghav-io commented on packages/core/getUserAvailability.ts:
```diff
@@ -324,7 +339,10 @@ const _getUserAvailability = async function getUsersWorkingHoursLifeTheUniverseA
+  let schedule = !useHostSchedulesForTeamEvent && eventType?.schedule ? eventType.schedule : userSchedule;
+
+  const hostSchedule = eventType?.hosts?.find((host) => host.user.id === user.id)?.schedule;
+  schedule = useHostSchedulesForTeamEvent && hostSchedule ? hostSchedule : schedule;
```
i have updated to be concise, with 'undefined' tests were failing as schedule can be null, so used nullish Coalescing and verified tests are passing.

vijayraghav-io committed:
update for initial value

Amit91848 commented on apps/web/components/eventtype/EventAvailabilityTab.tsx:
```diff
@@ -278,25 +280,152 @@ const EventTypeSchedule = ({ eventType }: { eventType: EventTypeSetup }) => {
+  );
+};
+
 const UseCommonScheduleSettingsToggle = ({ eventType }: { eventType: EventTypeSetup }) => {
```
This is where you can access team members from eventType.team.members for avatar and label for my previous comment

vijayraghav-io replied:
Updated

Amit91848 commented on apps/web/modules/event-types/views/event-types-single-view.tsx:
```diff
@@ -338,7 +346,15 @@ const EventTypePage = (props: EventTypeSetupProps & { allActiveWorkflows?: Workf
       minimumBookingNotice: eventType.minimumBookingNotice,
       metadata,
-      hosts: eventType.hosts,
+      hosts: eventType.hosts.map((host) => {
+        const member = eventType.team?.members.find((mem) => mem.user.id === host.userId) || null;
+        return {
+          ...host,
+          avatar: member?.user?.avatarUrl,
+          label: member?.user?.name,
+          availability: null,
+        };
```
This won't be needed along with other places where you are updating host with avatar and label

vijayraghav-io replied:
Updated

Amit91848 reviewed (changes requested):
Thanks for such quick changes 🔥, left a few more suggestions @vijayraghav-io . Also once you are done please update PR description with latest changes screenshot / video, @ciaranha would need to approve the ux and design changes that were made

Amit91848 commented on apps/web/components/eventtype/EventAvailabilityTab.tsx:
```diff
@@ -278,25 +280,152 @@ const EventTypeSchedule = ({ eventType }: { eventType: EventTypeSetup }) => {
+      : option.value === schedules?.find((schedule) => schedule.isDefault)?.id
+  );
+
+  if (!scheduleId) {
+    setValue(`hosts.${index}.scheduleId`, value?.value || null, { shouldDirty: true });
+    setValue(`hosts.${index}.availability`, value || null, { shouldDirty: true });
+  }
+
+  if (!availability) {
+    setValue(`hosts.${index}.availability`, value || null, { shouldDirty: false });
+  }
```
You can also clean this up and use the value directly inside the Select.

vijayraghav-io replied:
scheduleId has to be initialized, else 'null' will be sent to backend when -> new host is added in Assignment Tab but user does not change the default schedule selection. As OnChange() triggers only if there is change in selection.

vijayraghav-io committed:
updated and cleanup

vijayraghav-io committed:
set initial value

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#15416_roundRobinAvailability

vijayraghav-io committed:
update after merge

vijayraghav-io commented:
Thanks @Amit91848 , Updated. 
Also description is updated with latest UI video.

vijayraghav-io committed:
updated nit style

Amit91848 commented on apps/web/components/eventtype/EventAvailabilityTab.tsx:
```diff
@@ -281,25 +285,177 @@ const EventTypeSchedule = ({ eventType }: { eventType: EventTypeSetup }) => {
+      : option.value === schedules?.find((schedule) => schedule.isDefault)?.id
+  );
+
+  if (!scheduleId && !getFieldState(`hosts.${index}.scheduleId`).isDirty) {
+    setValue(`hosts.${index}.scheduleId`, value?.value || null, { shouldDirty: true });
+  }
```
This is still sending null to backend if you add host and save directly without making changes in the availability tab? Not sure what this is achieving? Imo this is not required. In case no scheduleId is being selected, aren't we defaulting to users default schedule anyway?


https://github.com/user-attachments/assets/99174cc4-bce3-4dc5-a1fe-7d82d862bb09

vijayraghav-io replied:
This may occur if host is added in Assignment Tab and form is saved without visiting Availability Tab.
Alternatively, this can be moved to submit handler, but there we will not have AvailabilityOptions to set the default schedule id. As  now we are confined AvailabilityOptions to TeamMemberSchedule component only and removed in Host. And we cant rely on .team.member.user.defaultScheduleId as this not always populated. (had checked this in trpc endpoint).

Anyhow the functionality of getUserAvailability() will still pick the default Schedule, as this how its updated at backend. So the event booking will not break.

Will update to handle this scenario, but not sure if this line can be completely removed. Will check and update.

Amit91848 commented:
This is working for collective events too, do you want to limit it to round robin only @ciaranha ?

vijayraghav-io commented on apps/web/modules/event-types/views/event-types-single-view.tsx:
```diff
@@ -459,7 +465,9 @@ const EventTypePage = (props: EventTypeSetupProps & { allActiveWorkflows?: Workf
     ),
-    availability: <EventAvailabilityTab eventType={eventType} isTeamEvent={!!team} />,
+    availability: (
+      <EventAvailabilityTab eventType={eventType} isTeamEvent={!!team} teamMembers={teamMembers} />
```
The type is not defined for eventType.team.members and to avoid lint errors using any, TeamMembers under eventTypeSetupProps is used, which is used in other components as well and has the same required data.

vijayraghav-io commented on apps/web/components/eventtype/EventTeamTab.tsx:
```diff
@@ -310,6 +328,18 @@ const Hosts = ({
     );
   }, [schedulingType, setValue, getValues, submitCount]);
 
+  const updatedHosts = (changedHosts: Host[]) => {
+    return changedHosts.map((newValue) => {
+      const existingHost = getValues("hosts").find((host: Host) => host.userId === newValue.userId);
+      return existingHost
+        ? {
+            ...newValue,
+            scheduleId: existingHost.scheduleId,
+          }
+        : newValue;
+    });
+  };
+
```
Yes this is required, else whenever a new host of same type is added , already added (or existing) host will loose its scheduleId.
So this is done, to ensure previously added hosts do no loose scheduleId.
This is because the host is created from list option in `CheckedHostField` component.

Amit91848 replied:
Makes sense, please add a comment because I couldn't figure out why it was added just by the code.

Amit91848 replied:
```suggestion
  const updatedHosts = (changedHosts: Host[]) => {
    const existingHosts = getValues("hosts");
    return changedHosts.map((newValue) => {
      const existingHost = existingHosts.find((host: Host) => host.userId === newValue.userId);
      return existingHost
        ? {
            ...newValue,
            scheduleId: existingHost.scheduleId,
          }
        : newValue;
    });
  };

```

vijayraghav-io replied:
added comment and updated

vijayraghav-io commented on apps/web/components/eventtype/EventTeamTab.tsx:
```diff
@@ -330,7 +362,7 @@ const Hosts = ({
                 value={value}
                 onChange={(changeValue) => {
-                  onChange([...value.filter((host: Host) => !host.isFixed), ...changeValue]);
+                  onChange([...value.filter((host: Host) => !host.isFixed), ...updatedHosts(changeValue)]);
```
Explained above

vijayraghav-io commented on apps/web/components/eventtype/EventAvailabilityTab.tsx:
```diff
@@ -281,25 +285,177 @@ const EventTypeSchedule = ({ eventType }: { eventType: EventTypeSetup }) => {
+        </div>
+        <div className="border-subtle rounded-b-md border border-t-0 p-6">
+          {hosts.length > 0 ? (
+            <ul
```
I had added for <li >, will check

vijayraghav-io replied:
though key prop was added , warning was occuring due to empty parent at 398, removed it

vijayraghav-io committed:
update for console warning and added comments

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#15416_roundRobinAvailability

Amit91848 commented on apps/web/components/eventtype/EventAvailabilityTab.tsx:
```diff
@@ -281,25 +285,171 @@ const EventTypeSchedule = ({ eventType }: { eventType: EventTypeSetup }) => {
+        <p className="text-emphasis my-auto ms-3 text-sm">{label}</p>
+      </div>
+      <div className="flex w-full flex-col pt-2 ">
+        <Controller
+          name={`hosts.${index}.scheduleId`}
+          render={({ field }) => {
+            return (
+              <Select
+                placeholder={t("select")}
+                options={options}
+                isSearchable={false}
+                onChange={(selected) => {
+                  field.onChange(selected?.value || null);
+                }}
+                className="block w-full min-w-0 flex-1 rounded-sm text-sm"
+                value={value as AvailabilityOption}
+                components={{ Option, SingleValue }}
+                isMulti={false}
+                isDisabled={isPending}
+              />
+            );
+          }}
+        />
```
NIT: 
```suggestion
        {isPending ? (
          <Spinner className="mt-2 h-6 w-6" />
        ) : (
          <Controller
            name={`hosts.${index}.scheduleId`}
            render={({ field }) => {
              return (
                <Select
                  placeholder={t("select")}
                  options={options}
                  isSearchable={false}
                  onChange={(selected) => {
                    field.onChange(selected?.value || null);
                  }}
                  className="block w-full min-w-0 flex-1 rounded-sm text-sm"
                  value={value as AvailabilityOption}
                  components={{ Option, SingleValue }}
                  isMulti={false}
                  isDisabled={isPending}
                />
              );
            }}
          />
        )}
```

vijayraghav-io replied:
done

Amit91848 reviewed (dismissed):
You can use defualtScheduleId in teamMembers here, in most cases it would be there from my understanding (when user has completed the onboarding), otherwise we are falling back to default schedule anyways.

https://github.com/calcom/cal.com/blob/c68664d44284912c8d5c0327efa953712af7e6b4/apps/web/components/eventtype/EventTeamTab.tsx#L422-L424

Otherwise changes look good

Amit91848 commented:
@ciaranha can you confirm the designs. Also it works for both Round Robin and Collective, is this something we want or should this be restricted to Round Robin only ?

vijayraghav-io committed:
nit/chore

vijayraghav-io committed:
chore

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#15416_roundRobinAvailability

vijayraghav-io committed:
update for typechecks

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#15416_roundRobinAvailability

vijayraghav-io committed:
update for scheduleId param after merge

vijayraghav-io commented:
> @ciaranha can you confirm the designs. Also it works for both Round Robin and Collective, is this something we want or should this be restricted to Round Robin only ?

@ciaranha can you please confirm the designs / share your comments.

ciaranha commented:
> @ciaranha can you please confirm the designs / share your comments.

@vijayraghav-io thanks, this is great for now!

@itsberkaya I would say you could do a quick design pass at this tab and establish better spacing and clean it up a bit?
![CleanShot 2024-07-30 at 07 15 33@2x](https://github.com/user-attachments/assets/1bc0cad2-36b8-4e4b-a140-afb374363093)

vijayraghav-io commented:
> > @ciaranha can you please confirm the designs / share your comments.
> 
> @vijayraghav-io thanks, this is great start on the UX.
> 
> @itsberkaya Could do a design pass at this tab and clean it up a bit? ![CleanShot 2024-07-30 at 07 15 33@2x]

Thankyou @ciaranha , for your feedback.

vijayraghav-io commented:
Thankyou @ciaranha!

@itsberkaya can you please do the needful, here. Thankyou!

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#15416_roundRobinAvailability

vijayraghav-io commented:
Hi @Amit91848 / @ciaranha ,
Can you please approve from your ends, as this PR is functional (done few rounds of self-review, functional testing and also tested by Amit during reviews ) as per requirement and design also seems to be accepted except few cleanups.
Thank you!
cc : @keithwillcode

Amit91848 commented:
Hey @vijayraghav-io, sorry for the delay but we are having internal discussion and finalizing the specs for the feature. There are also some design concerns being looked into by itsberkaya. 

Also since this is a high risk feature, it will have to go through another round of reviews from some senior team member. This might take some while, please bear with us. We are trying to do it as quickly as possible.

vijayraghav-io commented:
Thankyou! @Amit91848 for the update.

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#15416_roundRobinAvailability

vijayraghav-io committed:
update after merge or new changes in metadata.config

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#15416_roundRobinAvailability

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#15416_roundRobinAvailability

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#15416_roundRobinAvailability

vijayraghav-io committed:
update after weights functionality

vijayraghav-io commented:
@ciaranha 
Updated for changes due to weights functionality.
Also updated the spacing in Availability tab, hope this is fine.
https://www.loom.com/share/b194ce55a54649edad65eb10587994c6?sid=3faf6346-e165-4e4d-adf4-a4a9f0193bb7

CarinaWolli commented on apps/web/public/static/locales/en/common.json:
```diff
@@ -2561,5 +2561,8 @@
   "number_of_options":"{{count}} options",
   "reschedule_with_same_round_robin_host_title": "Reschedule with same Round-Robin host",
   "reschedule_with_same_round_robin_host_description": "Rescheduled events will be assigned to the same host as initially scheduled",
+  "choose_hosts_schedule" : "Choose hosts schedule",
```
```suggestion
  "choose_hosts_schedule" : "Choose host schedules",
```
I think that's better

vijayraghav-io replied:
updated

CarinaWolli commented on apps/web/public/static/locales/en/common.json:
```diff
@@ -2561,5 +2561,8 @@
   "reschedule_with_same_round_robin_host_title": "Reschedule with same Round-Robin host",
   "reschedule_with_same_round_robin_host_description": "Rescheduled events will be assigned to the same host as initially scheduled",
+  "choose_hosts_schedule" : "Choose hosts schedule",
+  "hosts_schedule_description": "You can select different schedule for each host. Each host's default schedule is choosen by default.",
```
```suggestion
  "hosts_schedule_description": "You can select a different schedule for each host. Each host's default schedule is chosen by default.",
```

vijayraghav-io replied:
updated

CarinaWolli commented on packages/prisma/schema.prisma:
```diff
@@ -614,6 +617,7 @@ model Schedule {
   name         String
   timeZone     String?
   availability Availability[]
+  Host         Host[]
```
```suggestion
  host         Host[]
```

vijayraghav-io replied:
Thankyou! @CarinaWolli for the review, i have updated suggested changes in single commit

CarinaWolli reviewed (commented):
@itsberkaya do we already have designs for that? 

If designs are not yet ready I think it can be merged like that and we can do a follow-up with UI improvements

CarinaWolli commented:
@emrysal could you maybe review as well (high risk PR)

vijayraghav-io committed:
updated description text and column name as per review suggestion

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#15416_roundRobinAvailability

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#15416_roundRobinAvailability

vijayraghav-io committed:
resolved merge conflicts

vijayraghav-io committed:
fix typecheck due to changes in getScheduleSchema

vijayraghav-io committed:
update after merge having refactor

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#15416_roundRobinAvailability

olivierlambert commented:
Gentle reminder that this PR is a pretty cool one to be merged, fixing some issues ;)

vijayraghav-io commented:
@Amit91848 , can you please take this PR forward 🙏

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#15416_roundRobinAvailability

vijayraghav-io committed:
update refactored EventTypeWebWrapper after merge

vijayraghav-io committed:
updated props

Amit91848 commented:
Hey @vijayraghav-io sorry for the delay. Code looks good as far as I am concerned, but will take a look again tomorrow. But this being high risk, will have to wait for someone from foundation team to review and approve it. We have bumped up the priority but they are already swamped so it can take a while.

vijayraghav-io commented:
> Hey @vijayraghav-io sorry for the delay. Code looks good as far as I am concerned, but will take a look again tomorrow. This pr being high risk, will have to wait for someone from foundation team to review and approve it. We have bumped up the priority but they are already swamped so it can take some time.

Thanks @Amit91848 for update, hoping to happen soon 🙏

emrysal commented on packages/trpc/server/routers/viewer/availability/schedule/getAllSchedulesByUserId.handler.ts:
```diff
@@ -0,0 +1,58 @@
+  });
+
+  if (!schedules) {
+    throw new TRPCError({
```
Is the best thing to do here unauthorized? It's a situation that should never happen (but could happen) - but not sure that UNAUTHORIZED is the best approach.

vijayraghav-io replied:
Thankyou! @emrysal for starting to review 🙏 

yeah may be is "NOT_FOUND" more apt here? Please let me know if this is ok.
Assuming that there will be at least one schedule for a given user, this may never happen, may be i added in case to handle the error if for some reason Prisma is not able to fetch the data.

vijayraghav-io replied:
it's a error message for very rarely occurring scenario we can update.

emrysal replied:
NOT_FOUND is good, also if you can add a logger to log this? We recently increased our log verbosity settings so it makes sense to log it using the logger.

vijayraghav-io replied:
Thankyou! @emrysal  for confirming, 
updated error code to NOT_FOUND, also added log.

vijayraghav-io replied:
@emrysal, can you please let know if you have any other comments or it is ok.

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#15416_roundRobinAvailability

vijayraghav-io committed:
updated trpc err code and added log

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#15416_roundRobinAvailability

anikdhabal commented:
Hey @vijayraghav-io could you pls fix the conflicts?

vijayraghav-io commented:
> Hey @vijayraghav-io could you pls fix the conflicts?

Sure, will update.

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#15416_roundRobinAvailability

vijayraghav-io committed:
update due to refactor in EventAvailabilityTab

vijayraghav-io commented:
@anikdhabal , resolved merge conflicts.

alishaz-polymath committed:
Merge branch 'main' into feat_#15416_roundRobinAvailability

emrysal committed:
Merge branch 'main' into feat_#15416_roundRobinAvailability

emrysal commented:
/tip 25 for patience whilst we get through our PR backlog

emrysal committed:
Merge branch 'main' into feat_#15416_roundRobinAvailability

emrysal committed:
perf: Optimise TS type + select only the data that's needed for a specific function

emrysal committed:
Merge branch 'main' into feat_#15416_roundRobinAvailability

emrysal reviewed (approved):
Tested this thoroughly on local, all tests passing and happy to merge. 👏

vijayraghav-io commented:
> Tested this thoroughly on local, all tests passing and happy to merge. 👏

Thankyou! @emrysal  🙏

emrysal commented:
/tip 25 @vijayraghav-io for patience whilst we get through our PR backlog

emrysal commented:
/tip $25 @vijayraghav-io for patience whilst we get through our PR backlog
